### PR TITLE
Added notification support to OPC-UA driver

### LINKED ISSUE #1934: Add notification support to OPC-UA driver
The OPC-UA Driver can be extended to include support for OPC-UA subscriptions and notifications using the `ChannelListener` APIs.

### PR DESCRIPTION
**Pull request by nicolatimeus**
Closes #1934

Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

### CODE REVIEW

**Comment by kartben** [comment]
I know this will sound like a "why does the f*** is he bothering us with that" kind of comment, and I want to apologize for that, but it would really be nice to see more code comments in the Kura codebase - and unfortunately this PR seems to be an example :(
With Kura Wires and the notion of Drivers being pretty new, I think it is safe to assume that people will want to use existing "substantial" drivers as reference implementations for creating their own (at least I know that's what I've tried to do) drivers, so it would probably be helpful to provide a bit more context wrt what class does what, etc.

**Comment by nicolatimeus** [comment]
Yeah sorry about that...
I agree that drivers like this can look a bit scary for people approaching Drivers for the first time, especially if there are no comments. 
Maybe we could consider adding a simpler "dummy" driver implementation under the examples folder that does nothing really useful apart from showing the basic steps involved in writing a Driver. This could also be used as a starting point for quickly creating a new Driver, without having to pick a "real" one and remove all the protocol-specific stuff.

**Comment by kartben** [comment]
A dummy driver would be a great idea!

File: kura/org.eclipse.kura.driver.opcua.provider/src/main/java/org/eclipse/kura/internal/driver/opcua/OpcUaChannelDescriptor.java [hunk 1]
```diff
@@ -132,4 +189,24 @@ static NodeId getNodeId(Map<String, Object> properties, int nodeNamespaceIndex,
...
         }
     }
-}
+
+    public static AttributeId getAttributeId(Map<String, Object> properties) {
+        String attributeId = properties.get(ATTRIBUTE_PROP_NAME).toString();
+        return AttributeId.valueOf(attributeId);
+    }
+
+    public static Double getSamplingInterval(Map<String, Object> properties) {
```
Comment (amitjoy) [inline kura/org.eclipse.kura.driver.opcua.provider/src/main/java/org/eclipse/kura/internal/driver/opcua/OpcUaChannelDescriptor.java]: could we probably use the primitive one?

File: kura/org.eclipse.kura.driver.opcua.localization/src/main/java/org/eclipse/kura/driver/opcua/localization/OpcUaMessages.java [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright (c) 2016, 2017 Eurotech and/or its affiliates and others
+ * Copyright (c) 2016, 2018 Eurotech and/or its affiliates and others
```
Comment (MMaiero) [inline kura/org.eclipse.kura.driver.opcua.localization/src/main/java/org/eclipse/kura/driver/opcua/localization/OpcUaMessages.java]: @nicolatimeus Let's remove localisation as this causes issues.

File: kura/org.eclipse.kura.driver.opcua.provider/OSGI-INF/metatype/org.eclipse.kura.driver.opcua.xml [hunk 3]
```diff
@@ -78,6 +78,15 @@
 			default="urn:kura:opcua:client"
 			description="OPC-UA application uri" />
 			
+		<AD id="subscription.publish.interval" 
+			name="subscription.publish.interval"
+			type="Long" 
+			cardinality="0" 
+			required="true" 
+			default="1000"
+			description="The publish interval in milliseconds for the subscripton created by the driver.">
```
Comment (MMaiero) [inline kura/org.eclipse.kura.driver.opcua.provider/OSGI-INF/metatype/org.eclipse.kura.driver.opcua.xml]: misspelled subscription

File: kura/org.eclipse.kura.driver.opcua.provider/src/main/java/org/eclipse/kura/internal/driver/opcua/OpcUaDriver.java [hunk 4]
```diff
@@ -398,150 +194,79 @@ public void unregisterChannelListener(final ChannelListener listener) throws Con
...
-            try {
-                opcuaType = OpcUaChannelDescriptor.getOpcuaType(channelConfig);
-            } catch (final Exception e) {
-                fail(record, message.errorRetrievingOpcuaType());
-                return Optional.empty();
-            }
-
+    private synchronized void onFailure(final ConnectionManager manager, final Throwable ex) {
+        if (connectionManager.isPresent() && connectionManager.get() == manager) {
+            logger.debug("Unrecoverable failure, forcing disconnect: {}", ex);
```
Comment (markoer) [inline kura/org.eclipse.kura.driver.opcua.provider/src/main/java/org/eclipse/kura/internal/driver/opcua/OpcUaDriver.java]: Exception needs no placeholder in the message. Remove the ': {}'.

File: kura/org.eclipse.kura.driver.opcua.provider/src/main/java/org/eclipse/kura/internal/driver/opcua/OpcUaDriver.java [hunk 5]
```diff
@@ -398,150 +194,79 @@ public void unregisterChannelListener(final ChannelListener listener) throws Con
...
-                fail(record, message.errorRetrievingNodeIdType());
-                return Optional.empty();
+                if (connectionManager.isPresent()) {
+                    this.disconnect();
+                }
+            } catch (ConnectionException e) {
+                logger.warn("Unable to Disconnect...");
             }
+        } else {
+            logger.debug("Ignoring failure from old connection: {}", ex);
```
Comment (markoer) [inline kura/org.eclipse.kura.driver.opcua.provider/src/main/java/org/eclipse/kura/internal/driver/opcua/OpcUaDriver.java]: Again.

**Review by markoer — COMMENTED** [review]
Can't really comment anything other than the two items below and the fact that some new tests would be nice, even without code comments.

File: kura/org.eclipse.kura.driver.opcua.provider/src/main/java/org/eclipse/kura/internal/driver/opcua/DataTypeMapper.java [hunk 6]
```diff
@@ -43,72 +43,121 @@ private static Variant mapByteArray(byte[] javaValue, VariableType opcuaType) {
...
                 return new Variant(stringValue);
             }
         }
-        throw new IllegalArgumentException(
-                message.errorValueTypeConversion() + javaValue.getClass() + " " + opcuaType.name());
+        throw new IllegalArgumentException("Error while converting the retrieved value to the defined typed "
+                + value.getClass() + " " + targetType.name());
+    }
+
+    private static byte[] toByteArray(Object calue) {
```
Comment (MMaiero) [inline kura/org.eclipse.kura.driver.opcua.provider/src/main/java/org/eclipse/kura/internal/driver/opcua/DataTypeMapper.java]: value?

Comment (markoer) [reply kura/org.eclipse.kura.driver.opcua.provider/src/main/java/org/eclipse/kura/internal/driver/opcua/DataTypeMapper.java]: I saw it too, but there's `value` a couple lines down as well... Would be interested to know what `calue` stands for.

File: kura/org.eclipse.kura.driver.opcua.provider/src/main/java/org/eclipse/kura/internal/driver/opcua/SubscriptionManager.java [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+        }
+
+        @Override
+        public CompletableFuture<Void> subscribe() {
+            return completedFuture(null);
+        }
+
+        @Override
+        public CompletableFuture<Void> unsubscribe() {
+            logger.debug("Unsubscrbing..");
```
Comment (MMaiero) [inline kura/org.eclipse.kura.driver.opcua.provider/src/main/java/org/eclipse/kura/internal/driver/opcua/SubscriptionManager.java]: unsubscribing

**Review by MMaiero — APPROVED** [review]

**Comment by MMaiero** [comment]
I believe it could be merged. Eventual small fixes will be done later in the release branch

**Pull request merged**